New module: sigprofiler

kdavydzenka opened the pull request:
## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

LouisLeNezet commented on modules/nf-core/sigprofiler/main.nf:
```diff
@@ -0,0 +1,38 @@
+   
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
```suggestion
    def args   = task.ext.args   ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
```

LouisLeNezet commented on modules/nf-core/sigprofiler/main.nf:
```diff
@@ -0,0 +1,38 @@
+    tuple val(meta), path(tsv_list, stageAs: '*.tsv')
+    
+    output:
+    tuple val(meta), path("results/*"),    emit: sigprofiler_results
+    path "versions.yml",                   emit: versions
```
```suggestion
    tuple val(meta), path("results/*"), emit: sigprofiler_results
    path "versions.yml",                emit: versions
```

LouisLeNezet commented on modules/nf-core/sigprofiler/main.nf:
```diff
@@ -0,0 +1,38 @@
+
+    """
+    mkdir -p results
+    echo "stub output" > results/dummy.txt
```
Instead of a dummy.txt, could you create file with the same names that would be created ?
```suggestion
    touch results/dummy.txt
```

LouisLeNezet commented on modules/nf-core/sigprofiler/templates/main_script.py:
```diff
@@ -0,0 +1,177 @@
+
+opt = dict()
+opt["prefix"] = "${task.ext.prefix ?: meta.id}"
+opt["dataset"] = "${task.ext.dataset ?: meta.dataset}"
```
If the dataset is always expected to be provided, can you add it as a value channel and add it to the meta.yml ?

kdavydzenka replied:
The dataset is always expected to be provided, and it has already been included in the meta.yml via the `meta` map.

LouisLeNezet reviewed (changes requested):
Hi,
I'm not yet quite familiar on python modules but here are some comment !

Katerina10-cloud commented on tests/config/nf-test.config:
```diff
@@ -47,6 +47,7 @@ profiles {
     singularity {
         singularity.enabled     = true
         singularity.autoMounts  = true
+        singularity.runOptions  = '--writable-tmpfs'
```
Thank you for pointing that out. Initially, I added `singularity.runOptions = '--writable-tmpfs'` because I encountered an issue when testing the` sigprofiler `tool, specifically when` make_decomposition_plots = true`. That option required a writable` /tmp` inside the container. However, I've adjusted the setup to use `make_decomposition_plots = false`, which avoids the need for a writable `/tmp.` So this change is no longer necessary, and I removed it from this PR

Katerina10-cloud commented on modules/nf-core/sigprofiler/templates/main_script.py:
```diff
@@ -0,0 +1,177 @@
+    subprocess.run(generate_count_matrix, shell=True)
+
+    matrix_files = {
+            "SBS96": os.path.join("output", "SBS", f"{dataset_id}.SBS96.all"),
+            "DBS78": os.path.join("output", "DBS", f"{dataset_id}.DBS78.all"),
+            "ID83": os.path.join("output", "ID", f"{dataset_id}.ID83.all")
```
Yes, these files are expected to be generated once per dataset for each genome.

kdavydzenka commented on modules/nf-core/sigprofiler/tests/main.nf.test:
```diff
@@ -0,0 +1,76 @@
+    tag "modules"
+    tag "modules_nfcore"
+    tag "sigprofiler"
+
```
Unfortunately, it's not feasible to add this. This is because the required genome data is specific to the tool and too large (~3 GB) to be published in the `nf-core/test-datasets`, even in a subsampled form (the tool requires all genome .txt files to be fully installed), which makes a minimal test case impractical.
I was able to run this test successfully in my local environment to verify functionality.

SPPearce commented:
It looks to me like this `dataset` is only being used in the code to set the name of the output files (and a column inside them) - this is what we generally use `prefix` for, setting by default to be `meta.id`.

itrujnara commented on modules/nf-core/sigprofiler/main.nf:
```diff
@@ -0,0 +1,38 @@
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        sigprofilermatrixGenerator: \$(python3 -c "import SigProfilerMatrixGenerator as matGen; print(matGen.__version__)")
```
You should also output Python version

itrujnara replied:
This also applies to the versions from the Python template script

itrujnara commented on modules/nf-core/sigprofiler/tests/main.nf.test:
```diff
@@ -0,0 +1,80 @@
+                    [ id:'test' ], 
+                    file(params.modules_testdata_base_path + 'delete_me/sigprofiler/test_GRCh38/HCC1395_WXS_HCC1395_joint_table.tsv', checkIfExists: true),
+                ] 
+                input[1] = []
```
Can we get a test where this option is used?

kdavydzenka replied:
> Can we get a test where this option is used?

Unfortunately, it's not feasible to add this test: `genome_installed_path and download_genome_sigprofiler: False `. This is because the required genome data is specific to the tool and too large (~3 GB) to be published in the `nf-core/test-datasets`, even in a subsampled form (the tool requires all genome .txt files to be fully installed), which makes a minimal test case impractical.
I was able to run this test successfully in my local environment to verify functionality.

itrujnara reviewed (commented):
Hi, the prefix handling looks good now, thanks! I've left some comments on other issues to address

kdavydzenka commented:
Dear @LouisLeNezet and @itrujnara,
I’ve addressed all your latest comments and made some minor updates to the module. 
Please let me know if anything else needs attention.

LouisLeNezet commented on modules/nf-core/sigprofiler/tests/main.nf.test:
```diff
@@ -0,0 +1,80 @@
+                    path(process.out.versions[0]).yaml
+                ).match() },
+                {
+                    // Directory existence check
+                    resultDirs.each { dir ->
+                        def file = new File(dir.toString())
+                        assert file.exists() 
+                        
+                    }  
```
Better than just checking if the files exist could you add a recursive file(it).name in the snapshot ?
This would help to diagnose if the prefix are correctly use and if the file are properly named.
Using this snapshot you will also be able to complete the stub test to be more representative of a real processing

kdavydzenka replied:
I’ve updated the test so that it checks for the existence of a small, representative set of files for each detected context type. The tool produces a large number of subfolders and files in the `results/* `directory, but the exact structure depends on which mutational contexts are present in the input dataset. Now test dynamically detects the contexts produced and verifies the presence of representative files for each one.

LouisLeNezet commented on modules/nf-core/sigprofiler/templates/main_script.py:
```diff
@@ -0,0 +1,186 @@
+if __name__ == '__main__':
+
+    prefix = opt["prefix"]
+    genome = opt["genome"]
```
What happen if no genome is provided ?
If this value is mandatory I would add it as a value channel and document it in the meta.yml.
Same for the context_types.

kdavydzenka replied:
@LouisLeNezet thank you for the suggestion. The `genome `argument is indeed mandatory for this process to run correctly. I have added genome as a value channel in the input and documented it in meta.yml as well, ensuring it is always provided. 
Regarding `context_types`, these are optional and provided as default arguments to the tool. If not explicitly specified, the tool automatically runs on all mutational contexts detected in the input data.

LouisLeNezet reviewed (commented):
Thanks for adressing the different comment.
I'm not quite familiar with python processes yet.

I have a few comment left.

Could you also add tests that would throw the different errors ?

kdavydzenka commented:
> Thanks for adressing the different comment. I'm not quite familiar with python processes yet.
> 
> I have a few comment left.
> 
> Could you also add tests that would throw the different errors ?

Hi @LouisLeNezet,
I’ve addressed your latest suggestions and added a couple of tests that trigger the different error cases as well. However, I wasn’t able to find clear examples in the nf-core documentation on the best way to write asserts for these error scenarios.

kdavydzenka commented on modules/nf-core/sigprofiler/main.nf:
```diff
@@ -0,0 +1,49 @@
+    
+    mkdir -p results/DBS78/DBS78/Suggested_Solution/COSMIC_DBS78_Decomposed_Solution/Signatures/
+    touch results/DBS78/DBS78/Samples.txt
+    touch results/DBS78/DBS78/Suggested_Solution/COSMIC_DBS78_Decomposed_Solution/Signatures/COSMIC_DBS78_Signatures.txt
```
Hi @maxulysse and @SPPearce.
Thanks for pointing this out! 
I’ve reworked the stub section so that paths and filenames are derived dynamically from the input `context_type `parameter. 
Regarding the repeated/nested paths, the directory and filenames are dictated by `SigProfiler` tool itself as well as output results nested structure.
